Issue #121 - Uses Deprecated neon-animation - remove

homerjonathan opened the pull request:
## Deprecated neon-animation

Fixes #121 

Component currently uses the neon-animation component which shows at https://www.webcomponents.org/element/PolymerElements/neon-animation to be deprecated.  Neon-animation component page recommends that we use web animations API or CSS animations as a replacement. I have chosen the **CSS Animations** route.  

## Overview of Changes

Uses simple CSS animation.  Adds --paper-tooltip-delay so that delay can be set externally.  Uses updateStyles so that this value can still be set by properties.  However default works well and now allows this value to be styled.
 
## ES6

I have also taken the opportunity to upgrade the object to Polymer 2.0 ES6 rather than hybrid.

## Tests

Default Demo's continue to work.

homerjonathan commented:
Ive signed my life away ;-)

notwaldorf commented:
cc @e111077 
My understanding is that this would be a breaking change to the element, and if we merge it we'd have to bump the major version.

homerjonathan commented:
You are right in that it changes a number of things.  I automatically update to full ES6 Polymer 2.0 as a matter of course.  I still have hybrid mode components and the changes has caused no issues and required no changes to the code.

Since Neon-Animation uses the following structure:

```javascript
  animationConfig: {
          type: Object,
          value: function() {
            return {
              'entry': [{
                name: 'fade-in-animation',
                node: this,
                timing: {delay: 0}
              }],
              'exit': [{
                name: 'fade-out-animation',
                node: this
              }]
            }
          }
        },
```
Which is very specific to Neon-Animation, since this has been deprecated there is no alternative.  However I am guessing that most users will be just using the simple fade-in with using the mixin's to format in the right colours etc.  However it does open up the range of CSS Animations which seems very powerful so you can add more interesting effects.  Checking https://caniuse.com/#search=CSS%20Animation it seems that it is very well supported so should limit impact of the change.  

I am going to investigate the **continuous-integration/travis-ci/pr — The Travis CI build failed** not sure why this has gone wrong.  However this is my first public pull request so be kind! ;-)

notwaldorf commented:
Oh! Congrats on your first pull request! We try to be kind no matter what, but we'll be extra on this one :)

I'm afraid the reason why I was mentioning that it was a breaking change is that it's unlikely this PR can get merged at this time -- especially because, as you mentioned, you are making this element not-hybrid, we don't have bandwidth to maintain it in this form. We wrote a bit on that topic here: https://www.polymer-project.org/blog/2017-11-27-future-of-elements.html

homerjonathan commented:
Thanks for the link.  It's trivial to update from Hybrid to Class mode, however I can understand why your holding back on the ES6 until Polymer 3.0.  

Most of the changes removing Neon-Animation would be easy to to retrofit back to the Hybrid Mode.  Give me a few days...I will drag it back to a Hybrid version.

notwaldorf commented:
I'll let @e111077 chime in, since he owns all neon-animations things, but just removing neon-animations is still a breaking change, since the API is changing. Is this something we will merge, Elliott?

e111077 commented:
Monica is right that simply removing `animationConfig` is a straight-up breaking change, but I do suspect that this will not really affect that many people. I'll have to test this internally soon.

Though, I still do not believe that we will be able to merge the ES6. as we have many many users that are still using Polymer 1 / hybrid mode.

homerjonathan commented:
So if I return this to a Hybrid version. Would that remove the road blocks for a merge?

e111077 commented:
I still have to run tests internally to see how much removing `animationConfig` affects existing projects, but returning to hybrid will remove some of our blockers and will definitely move this PR towards a mergeable state.

homerjonathan commented:
Commit converts back to Hybrid Mode.  Need to run tests but works in my application.  Will do the tests later and add any comments. etc.

homerjonathan commented:
Since Neon-Animation has been removed the finish animation event does not trigger.  This failed the **tooltip is hidden after target is blurred** test.  I have updated the test, watching for a "**blur**" event instead on the target of the tooltip.  Since the Polymer component also gets the same event, I have added a short sleep to maker sure the Polymer component has time to hide itself, stopping a possible chase condition error.  The test **isHidden** then successfully returns that the tooltip has been hidden and the test is passed.  

Not sure if the sleep is the best way of adding to the test.  However unsure of a neater solution.  

However its over to you guys for feedback, suggestions etc.

e111077 commented:
Again, nice work. I just wanted to update that this is still being actively looked at.

e111077 commented:
This PR breaks several YouTube projects that call `playAnimation` and `cancelAnimation`. Also animationConfig must have a type declared

homerjonathan commented:
playAnimation and cancelAnimation is fairly easy to add and link simply to show and hide. Adding animationConfig is easy too.... but to replicate NeonAnimation would be difficult. What particular animation does YouTube use? Perhaps we can replicate using CSS for particular requests, kind of like a simulator. We wouldn't want to break YouTube as someone might notice 🤣

homerjonathan commented:
Looking at the idea of a simulator for the NeonAnimation API. Most of the basic animations are easy to replicate in CSS. I was considering reading the animation config and simply choosing the right animation CSS class from their choice. If the animation config is fairly simple it should work. I would need YouTubes config to set the scope on how complex the choices might be. If the user wants even more complex animations they then can go down the mixin route as the final resort.

homerjonathan commented:
Looking at the API do they change the Animation Type? There is the fade-in-animation which is the most likely choice for tooltip. However there are others like slide up, down etc.  That and changing timing like delay would be easy to read the animationConfig and set the values etc. However if YouTube add more steps to the animationConfig that would be difficult to emulate.

animationConfig.entry[0].timing is the 'entry' animation. So any change to this could be simply routed through to change the animation timing.

e111077 commented:
It seems that this specific project uses the same animation. You'll have to give me several hours to run more-comprehensive tests to see if any other projects are affected and change the animation type.

homerjonathan commented:
Elliott don't forget to check if the exit animation is being used, animationConfig.entry[1].timing etc. It may be another thing that has been fiddled with. I will have a go at solving the first issues raised with playAnimation over weekend.

homerjonathan commented:
Here is an update that I think will reduce the issues.  First **playAnimation** is simply routed to functions **show** and **hide**.  I read the **animationConfig** object and select the CSS with that name to produce the animation effect.  I have created the CSS equivalent for _fade-in-animation_, _fade-out-animation_, _scale-up-animation_ and _scale-down-animation_.  It has the added bonus that if you want to create your own animation you can create it in CSS and it will run that instead.  I do need to read the timing from the **animationConfig** object to complete a basic full emulation of the Neon-Animation. 

I have updated the test to use my mixin's to stop the delayed setting of hidden attribute from breaking all the tests.  I have now four mixin's for testing giving much more finer control of timing.

Just noticed the **cancelAnimation** that needs to be added, which would be easy to fix.  However this is quite a change of direction from my last commit.  Just want to check if you think this is a good direction before investing in the extra time to close any remaining issues and close any holes in the emulation.

The only issues I can think of now is that the Multi-Animation would not be supported with this version.  I think it would be possible using the **animationend** to trigger the next CSS in a kind of loop.  I think that wouldn't be a good idea as Neon-Animation is being deprecated as it was going in the wrong direction.  Thus adding basic support will stop most of the users of the tooltip from having to do too much redevelopment but still support the Polymer mantra of #UseThePlatform encouraging the route of adding CSS animation as a better direction.

Again back to you guys for comments and directions.

e111077 commented on paper-tooltip.html:
```diff
@@ -81,331 +72,413 @@
+          },
+          /**
+           * Replaces Neon-Animation playAnimation - just calls show and hide.
+           * @Deprecated Use show and hide instead.
```
`@deprecated` instead of `@Deprecated`

e111077 commented on paper-tooltip.html:
```diff
@@ -81,331 +72,413 @@
+           * @Deprecated Use show and hide instead.
+           * @param {string=} type
+           */
+          playAnimation: function(type) {
```
Please add a `cancelAnimation` function. Some tests still failing because of this.

I believe it should have something along the lines of:

```js
cancelAnimation: function() {
  this.$.tooltip.classList.remove(this._getAnimationType('entry'));
  this.$.tooltip.classList.remove(this._getAnimationType('exit'));
},
```

I didn't have time to take a deep look at how this would affect internal state so you may want to iterate on my impl

homerjonathan commented:
I have the changes suggested by Elliott.  There was a **_animationCancel** which was renamed to **cancelAnimation**.  Fixed I hope the 2 spaces per tab and the small case deprecated.  

So the only thing that I can see that needs sorting is the other animations that may be requested and to read the animationConfig for any timings that may have been set.  However I would like to see if there are any other components that use Neon-Animation to see if we can make sure we can use the time invested in this and make it easier to remove it from them as well.

After that I would like to add a test for **cancelAnimation** and update documentation for the mixins.

homerjonathan commented:
I have updated the Documentation and added a test to confirm that the functions **playAnimation** and **cancelAnimation** work.

I still need to check animationConfig for timings and set those.  I was also considering since animationConfig linked with Neon-Animation has been deprecated, that two attribute should be added to change the values as below.

```js
<paper-tooltip for="demo" entry-animation="fade-in-animation" exit-animation="fade-out-animation">
```
What do you think?

Finally to close all the gaps I will need to create the CSS for the other types left in Neon-Animation.

- slide-down-animation Animates transform from none to translateY(100%);
- slide-up-animation Animates transform from none to translateY(-100%);
- slide-from-top-animation Animates transform from translateY(-100%) to none;
- slide-from-bottom-animation Animates transform from translateY(100%) to none;
- slide-left-animation Animates transform from none to translateX(-100%);
- slide-right-animation Animates transform from none to translateX(100%);
- slide-from-left-animation Animates transform from translateX(-100%) to none;
- slide-from-right-animation Animates transform from translateX(100%) to none;

e111077 commented:
Ran full-codebase internal tests. 13/2713 tests failing. Still investigating reason for failures. Suspect they might not be related to your project.

e111077 commented:
[Link to diff ignoring whitespace](https://github.com/PolymerElements/paper-tooltip/pull/126/files?w=1)

e111077 commented on paper-tooltip.html:
```diff
@@ -202,53 +268,69 @@
+        // Short-cut and cancel all animations and hide
+        this.$.tooltip.classList.remove(this._getAnimationType('entry'));
+        this.$.tooltip.classList.remove(this._getAnimationType('exit'));
+        this.$.tooltip.classList.add('hidden');
```
Cancelling animation should not necessarily hide the tooltip; it should just skip the animation

homerjonathan replied:
Not sure how that links with playAnimation which I thought was it's opposite number. Since the tooltip will most likely be triggered by mouse movement I figured that by calling playAnimation it would of been a programmatic way of opening the tooltip. Thus cancelAnimation would be it's counter and hide the tooltip. Have I misunderstood the reason why playAnimation exists?  By removing the 'entry' and 'exit' the tooltip will disappear anyhow. The hidden just removes from the page completely the tooltip so it doesn't cause page issues.

Your thoughts?

homerjonathan replied:
I suppose a better question is when cancelAnimation is called. Is the tooltip been asked to complete to fully show itself or completely hide itself? Or does it depend on when it's called? The same is easier with playAnimation where it is set by 'entry' or 'exit'.

e111077 replied:
The previous way it worked is that there would be an entry and exit animation. playAnimation would play this fade in or fade out. cancelAnimation would stop playing the animation and return it to it's pre-animation state.

so if it is in the middle of playing the entry animation, cancel animation would make it hidden. if it is in the middle of playing the exit animation, cancel animation would return it to being visible.

homerjonathan commented:
Added **cancelAnimation** as requested.  This jumps to the end of the animation each way either on the way to show or on the way to hide.  This should now pass all internal tests as well as travis.

I have added one extra demo which allows the user to send `playAnimation('entry')`for entry animation, `playAnimation('exit')` for the exit animation and finally `cancelAnimation()` to jump to the end.  All of it #UseThePlatform!  

The only thing left to do is to pick up any timing changes made in **animationConfig** and add the other neon-animation types.

What do you think of my earlier suggestion of using attributes to set the animation type rather than **animationConfig**?

Also on a rather more annoying note does anyone know why my github keeps using DudleyAH which is an internal company github account that I do work for rather than mine! Its failing the CLA test because of this!

homerjonathan commented:
I have added **slide-down-animation** this works fine.  However when scrolling up realised that I needed to build a **slide-down-animation-out** (that doesn't exist) to help the tooltip leave the screen in the correct manner.  It was easy to create but looking at the specifications:

> slide-from-top-animation Animates transform from translateY(-100%) to none;

It uses the -100% this does not work well since it takes the size of 100% of the container which means that the tooltip will move only a few pixels (ie its height).  So I am not sure how the Neon-Animation would have worked with the tooltip.  I do like the drop down tooltip and it follows the Material Design Specification for speed.  However this part of the solution feels awkward so I am not sure if I am going in the right direction with the other animation types that Neon-Animation provides.  

So whats your input?  Shall I stick just to the Fade In and Out Animation and Scale Up and Down.  Or did tooltip use only a tiny fraction of Neon-Animation and we should just close the book and stop the emulation here?

homerjonathan commented:
I have completed all the changes I think that should be added to the component.

```html
<paper-tooltip id="demo5_tooltip" for="demo5_icon1" animation-delay="500" animation-entry="scale-up-animation" animation-exit="scale-down-animation">
```
Two new attributes **animation-entry** and **animation-exit** that allows changes to the animation without needing _animationConfig_ effectively giving an easy way for users to stop using the deprecated library.

The update also reads the timing.delay and will update itself to comply.  However using the mixins makes again an easy way out of needing _animationConfig_

I have added two more demo's.  One demonstrating the new animation attributes.  The second the use of **playAnimation** and **cancelAnimation**.  

Hopefully that should minimise any breaks for the code.  With an easy way to fix if it does.  

What are your thoughts?

e111077 commented:
Additionally, can you rebase this pr to master?

homerjonathan commented:
I have squashed the commits and rebased.  However despite typing in many commands it still showing as being behind master.  I am unsure what has gone wrong.  Any assistance?

homerjonathan commented:
Git is nice until you want to do something a bit more complicated.  I think I have the merge working correctly now!

e111077 commented:
This looks good to me. I just ask you fix that little spacing nit mentioned above.

e111077 commented:
@notwaldorf should we handle this as a minor or major change? There are some missing features, but tests in Google don't seem to break because of behavior, only build issues.

homerjonathan commented:
Spacing altered.  

There are some missing features.  Mainly it depends on how far you want to replicate Neon-Animation.  I can add the whole set but that would add to the size.  I have for example my slide from top animation.  But not in any other direction!  

Looking at the big picture I am confident I can write a fake Neon-Animation that would almost completely simulate it.  However I note comments that Neon-Animation was deprecated because it was going in the wrong direction.  So a full simulation _may be encouraging users_ to continue down that route.

e111077 commented:
I believe that this PR covers like 95% of the use cases. I don't think there are many people out there that are actively manipulating the animation on paper-tooltip. This seemed evident to me when I saw the number of failing tests Google-wide

e111077 commented:
🎉🎉🎉🎉🎉🎉🎉🎉🎉🎉🎉🎉🎉🎉🎉🎉🎉🎉🎉
for now will merge. Will decide on versioning in later PR.

Thanks a million for the submission!!

🎉🎉🎉🎉🎉🎉🎉🎉🎉🎉🎉🎉🎉🎉🎉🎉🎉🎉🎉

krumware commented:
Wanted to say thanks on this!!!!
